### PR TITLE
jcaip/benchmark: add better testing/example support

### PR DESCRIPTION
**Pull request by jcaip**
- modified example to be more generic, updated Makefile
- using sklearn for performance metrics now 
- parameter server now compatible with lr scheduler
- removed unneeded process init code
- moved models into separate file. 
- add some logs to the docs (may remove later on)

### CODE REVIEW

File: distbelief/optim/downpour_sgd.py [hunk 1]
```diff
@@ -70,12 +69,11 @@ def step(self, closure=None):
...
 
-        # NB: loss.backward() should be called before this, so that we are sure that the .grad in the model params
-        # can be accessed. This should be always satisfied since we call loss.backward() right before step()
+        #get the lr
+        lr = self.param_groups[0]['lr']
         # keep track of accumulated gradients so that we can send 
         gradients = ravel_model_params(self.model, grads=True)
-        # this does self.accumulated_gradients+=-self.lr * gradients
-        self.accumulated_gradients.add_(gradients)
+        self.accumulated_gradients.add_(-lr, gradients)
```
Comment (rohan-varma) [inline distbelief/optim/downpour_sgd.py]: The server also does `tensor.add_(-lr, gradients)`, so having it here will cause the gradients to be scaled by the learning rate twice. I think we should have `self.accumulated_gradients.add_(gradients)` here so that the server can do whatever it wants with the gradients (initially just scaling by the LR, but later we can do different sorts of optimizers on the client/servers.

File: example/main.py [hunk 2]
```diff
@@ -1,163 +1,142 @@
...
     print('Finished Training')
 
 
-def evaluate(net, testloader, dataset_name):
-    if dataset_name == 'MNIST':
-        classes = (0, 1, 2, 3, 4, 5, 6, 7, 8, 9)
-    else:
-        classes = ('plane', 'car', 'bird', 'cat', 'deer', 'dog', 'frog', 'horse', 'ship', 'truck')
+def evaluate(net, testloader, args):
+    classes = ('plane', 'car', 'bird', 'cat', 'deer', 'dog', 'frog', 'horse', 'ship', 'truck')
```
Comment (rohan-varma) [inline example/main.py]: We should have the correct class names here for the MNIST dataset if that is the dataset we are training with as well.

File: example/main.py [hunk 3]
```diff
@@ -1,163 +1,142 @@
...
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description='Distbelief training example')
+    parser.add_argument('--batch-size', type=int, default=64, metavar='N', help='input batch size for training (default: 64)')
+    parser.add_argument('--test-batch-size', type=int, default=10000, metavar='N', help='input batch size for testing (default: 1000)')
+    parser.add_argument('--epochs', type=int, default=10, metavar='N', help='number of epochs to train (default: 10)')
+    parser.add_argument('--lr', type=float, default=0.1, metavar='LR', help='learning rate (default: 0.1)')
+    parser.add_argument('--freq', type=int, default=10, metavar='N', help='how often to send/pull grads (default: 10)')
+    parser.add_argument('--cuda', action='store_true', default=False, help='use CUDA for training')
+    parser.add_argument('--log-interval', type=int, default=10, metavar='N', help='how many batches to wait before logging training status')
+    parser.add_argument('--distributed', action='store_true', default=False, help='whether to use DownpourSGD or normal SGD')
```
Comment (rohan-varma) [inline example/main.py]: should we have `--distributed` on by default?

File: example/main.py [hunk 4]
```diff
@@ -1,163 +1,142 @@
...
     transform = transforms.Compose([
                 transforms.ToTensor(),
                 transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))
             ])
 
-    trainloader, testloader = get_dataset(args.dataset, transform)
-
-    net = Net(num_channels=1 if args.dataset=='MNIST' else 3)
+    trainloader, testloader = get_dataset(args, transform)
+    net = LeNet()
```
Comment (rohan-varma) [inline example/main.py]: I don't think the current implementation of `LeNet` in the models file will work with the MNIST dataset, since MNIST has 1 channel while the first conv step in LeNet appears to expect 3 channels. The previous code looked at the number of channels in the input dataset to get the dims right for both `MNIST` and `CIFAR`.

File: example/main.py (hunk 2 above)
Comment (jcaip) [reply example/main.py]: good idea will udpate

File: example/main.py (hunk 3 above)
Comment (jcaip) [reply example/main.py]: yes, but i didn't quite know the proper syntax for it

File: example/main.py (hunk 4 above)
Comment (jcaip) [reply example/main.py]: ah kk will change back

Comment (jcaip) [reply example/main.py]: should we just take out MNIST support? i feel like it's not as useful as cifar

File: example/main.py (hunk 3 above)
Comment (rohan-varma) [reply example/main.py]: probably `default=True`

File: example/main.py (hunk 4 above)
Comment (rohan-varma) [reply example/main.py]: I think it's worth it to support `MNIST` just as a quicker way of getting a benchmark. If you want, I can go ahead and add back the support once you merge.

File: distbelief/optim/downpour_sgd.py (hunk 1 above)
Comment (rohan-varma) [reply distbelief/optim/downpour_sgd.py]: It looks like the current `server.py` still scales by the LR: https://github.com/ucla-labx/distbelief/blob/db54a1ae9937d2f495f28f65e98abacdec82632d/distbelief/server.py#L42. Agreed with your approach, but if we only add raw gradients to the server then the best we can do on the server is probably like `SGD`

Comment (rohan-varma) [reply distbelief/optim/downpour_sgd.py]: Sounds good! Will this mean that the server and clients will always be running the same optimizer?
